src/util/query_exec_util.cpp

```diff
@@ -0,0 +1,249 @@
+            common::ManagedPointer<planner::CreateTablePlanNode>(
+                reinterpret_cast<planner::CreateTablePlanNode *>(out_plan.get())),
+            common::ManagedPointer(accessor), db_oid_);
+        break;
+      case network::QueryType::QUERY_DROP_INDEX:
+        // Drop index does not need execution of compiled query
+        status =
+            execution::sql::DDLExecutors::DropIndexExecutor(common::ManagedPointer<planner::AbstractPlanNode>(out_plan)
+                                                                .CastManagedPointerTo<planner::DropIndexPlanNode>(),
+                                                            common::ManagedPointer<catalog::CatalogAccessor>(accessor));
+        break;
+      case network::QueryType::QUERY_CREATE_INDEX:
+        status = execution::sql::DDLExecutors::CreateIndexExecutor(
+            common::ManagedPointer<planner::AbstractPlanNode>(out_plan)
+                .CastManagedPointerTo<planner::CreateIndexPlanNode>(),
+            common::ManagedPointer<catalog::CatalogAccessor>(accessor));
+        if (status) {
+          execution::exec::ExecutionSettings settings{};
+          if (CompileQuery(query, nullptr, nullptr, std::make_unique<optimizer::TrivialCostModel>(), settings)) {
+            ExecuteQuery(query, nullptr, nullptr, nullptr, settings);
+          }
+        }
+        break;
+      default:
+        NOISEPAGE_ASSERT(false, "Unsupported QueryExecUtil::ExecuteStatement");
+        break;
+    }
+  }
+
+  return status;
+}
+
+bool QueryExecUtil::CompileQuery(const std::string &statement,
+                                 common::ManagedPointer<std::vector<parser::ConstantValueExpression>> params,
+                                 common::ManagedPointer<std::vector<type::TypeId>> param_types,
+                                 std::unique_ptr<optimizer::AbstractCostModel> cost,
+                                 const execution::exec::ExecutionSettings &exec_settings) {
+  NOISEPAGE_ASSERT(txn_ != nullptr, "Requires BeginTransaction() or UseTransaction()");
+  auto txn = common::ManagedPointer<transaction::TransactionContext>(txn_);
+  auto accessor = catalog_->GetAccessor(txn, db_oid_, DISABLED);
+  auto result = PlanStatement(statement, params, param_types, std::move(cost));
+  if (!result.first || !result.second) {
+    return false;
+  }
+
+  const std::unique_ptr<planner::AbstractPlanNode> &out_plan = result.second;
+  NOISEPAGE_ASSERT(network::NetworkUtil::DMLQueryType(result.first->GetQueryType()), "ExecuteDML expects DML");
+  common::ManagedPointer<planner::OutputSchema> schema = out_plan->GetOutputSchema();
+
+  auto exec_query = execution::compiler::CompilationContext::Compile(*out_plan, exec_settings, accessor.get(),
+                                                                     execution::compiler::CompilationMode::OneShot);
+  schemas_[statement] = schema->Copy();
+  exec_queries_[statement] = std::move(exec_query);
```

Comment:
Do we actually use this query cache? It doesn't seem that we're checking whether this query is cached before recompiling it.

Reply:
Thanks for the catch!